From cca0a11ca63582f9f03aaab2506e60848d54604f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Rafa=C5=82=20Dzi=C4=99giel?= <rafostar.github@gmail.com>
Date: Wed, 20 Jan 2021 11:46:17 +0100
Subject: [PATCH 1/2] assrender: Fix mimetype detection

Previously gst_structure_has_name was used to get a string to compare with supported mimetypes.
This is incorrect as above function returns a user defined structure name which is
not the structure mimetype value.
---
 ext/assrender/gstassrender.c | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/ext/assrender/gstassrender.c b/ext/assrender/gstassrender.c
index e6d31985b6..dbbf6ce245 100644
--- a/ext/assrender/gstassrender.c
+++ b/ext/assrender/gstassrender.c
@@ -1557,7 +1557,7 @@ gst_ass_render_handle_tag_sample (GstAssRender * render, GstSample * sample)
   const GstStructure *structure;
   gboolean valid_mimetype, valid_extension;
   guint i;
-  const gchar *filename;
+  const gchar *mimetype, *filename;
 
   buf = gst_sample_get_buffer (sample);
   structure = gst_sample_get_info (sample);
@@ -1568,10 +1568,13 @@ gst_ass_render_handle_tag_sample (GstAssRender * render, GstSample * sample)
   valid_mimetype = FALSE;
   valid_extension = FALSE;
 
-  for (i = 0; i < G_N_ELEMENTS (mimetypes); i++) {
-    if (gst_structure_has_name (structure, mimetypes[i])) {
-      valid_mimetype = TRUE;
-      break;
+  mimetype = gst_structure_get_string (structure, "mimetype");
+  if (mimetype) {
+    for (i = 0; i < G_N_ELEMENTS (mimetypes); i++) {
+      if (strcmp (mimetype, mimetypes[i]) == 0) {
+        valid_mimetype = TRUE;
+        break;
+      }
     }
   }
 
-- 
GitLab


From ec2b1e69099e0f8b7821dc217b21a29be10c9ab5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Rafa=C5=82=20Dzi=C4=99giel?= <rafostar.github@gmail.com>
Date: Wed, 20 Jan 2021 13:38:03 +0100
Subject: [PATCH 2/2] assrender: Do not iterate over mimetypes without filename

No point spending time on iterating and comparing strings if we
are going to reject the value due to missing filename anyway.
---
 ext/assrender/gstassrender.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/ext/assrender/gstassrender.c b/ext/assrender/gstassrender.c
index dbbf6ce245..a69d3fe789 100644
--- a/ext/assrender/gstassrender.c
+++ b/ext/assrender/gstassrender.c
@@ -1565,6 +1565,10 @@ gst_ass_render_handle_tag_sample (GstAssRender * render, GstSample * sample)
   if (!buf || !structure)
     return;
 
+  filename = gst_structure_get_string (structure, "filename");
+  if (!filename)
+    return;
+
   valid_mimetype = FALSE;
   valid_extension = FALSE;
 
@@ -1578,10 +1582,6 @@ gst_ass_render_handle_tag_sample (GstAssRender * render, GstSample * sample)
     }
   }
 
-  filename = gst_structure_get_string (structure, "filename");
-  if (!filename)
-    return;
-
   if (!valid_mimetype) {
     guint len = strlen (filename);
     const gchar *extension = filename + len - 4;
-- 
GitLab

